File: commands/stats/warzone.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone, timedelta
from collections import Counter
from database import db

class Warzone(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='warzone', description='Live 6-hour battle royale leaderboard')
    async def warzone(self, interaction: discord.Interaction):
        try:
            await interaction.response.defer()
            
            cutoff_time = datetime.now(timezone.utc) - timedelta(hours=6)
            query = {'timestamp': {'$gte': cutoff_time}}
            
            wins = []
            async for win in db.db.clanwins.find(query):
                wins.append(win)
            
            if not wins:
                embed = discord.Embed(
                    title="Warzone - No Activity",
                    description="No wins in the last 6 hours. The battlefield is quiet...",
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            clan_counter = Counter(win.get('clan_name', 'Unknown') for win in wins if win.get('clan_name'))
            top_clans = clan_counter.most_common(10)
            
            if not top_clans:
                embed = discord.Embed(
                    title="Warzone - No Activity",
                    description="No valid clan data in the last 6 hours.",
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            champion_clan, champion_wins = top_clans[0]
            
            clan_points = {}
            for clan_name, _ in top_clans:
                clan_wins_list = [w for w in wins if w.get('clan_name') == clan_name]
                clan_points[clan_name] = sum(w.get('player_count', 0) for w in clan_wins_list)
            
            now = datetime.now(timezone.utc)
            hours_since_start = now.hour % 6
            minutes_since_start = now.minute
            hours_until_reset = 5 - hours_since_start
            minutes_until_reset = 60 - minutes_since_start
            
            if minutes_until_reset == 60:
                hours_until_reset += 1
                minutes_until_reset = 0
            
            embed = discord.Embed(
                title="⚔️ WARZONE - Live Battle Royale",
                description=f"6-Hour Leaderboard | Resets in **{hours_until_reset}h {minutes_until_reset}m**",
                color=0xff4500
            )
            
            champion_points = clan_points.get(champion_clan, 0)
            embed.add_field(
                name="👑 KING OF THE HILL",
                value=f"**[{champion_clan}]**\n{champion_wins} wins | {champion_points} points",
                inline=False
            )
            
            rankings = []
            for i, (clan, count) in enumerate(top_clans, 1):
                points = clan_points.get(clan, 0)
                if i == 1:
                    medal = "👑"
                elif i == 2:
                    medal = "🥈"
                elif i == 3:
                    medal = "🥉"
                else:
                    medal = f"#{i}"
                
                rankings.append(f"{medal} [{clan}] - {count} wins | {points}pts")
            
            embed.add_field(
                name="🏆 Live Rankings",
                value="\n".join(rankings),
                inline=False
            )
            
            total_wins = len(wins)
            total_points = sum(w.get('player_count', 0) for w in wins)
            contest_wins = sum(1 for w in wins if w.get('contest', False))
            
            stats_text = f"Total Battles: **{total_wins}**\n"
            stats_text += f"Contest Battles: **{contest_wins}**\n"
            stats_text += f"Total Points: **{total_points}**\n"
            stats_text += f"Active Clans: **{len(top_clans)}**"
            
            embed.add_field(
                name="📊 Warzone Stats",
                value=stats_text,
                inline=False
            )
            
            latest_win = sorted(wins, key=lambda x: x.get('timestamp', datetime.min.replace(tzinfo=timezone.utc)), reverse=True)[0]
            win_time = latest_win.get('timestamp', datetime.now(timezone.utc))
            if win_time.tzinfo is None:
                win_time = win_time.replace(tzinfo=timezone.utc)
            
            time_ago = datetime.now(timezone.utc) - win_time
            mins_ago = int(time_ago.total_seconds() / 60)
            
            if mins_ago < 1:
                time_str = "just now"
            elif mins_ago < 60:
                time_str = f"{mins_ago}m ago"
            else:
                hours_ago = mins_ago // 60
                time_str = f"{hours_ago}h ago"
            
            embed.add_field(
                name="⚡ Latest Battle",
                value=f"[{latest_win.get('clan_name', 'Unknown')}] won on {latest_win.get('map', 'Unknown')} - {time_str}",
                inline=False
            )
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in warzone: {e}")
            await interaction.followup.send(embed=discord.Embed(
                title="Error", description="Failed to load warzone.", color=0xff0000
            ))

async def setup(bot):
    await bot.add_cog(Warzone(bot))

File: commands/stats/warzone.js
// https://github.com/viktorexe/terristats-discord-bot
// /warzone - a live 6-hour "battle royale" leaderboard of the most active
// clans, with a countdown to the next 6-hour reset.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('warzone')
    .setDescription('Live 6-hour battle royale leaderboard'),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const cutoff = new Date(Date.now() - 6 * 3600 * 1000);
      const wins = await db
        .collection('clanwins')
        .find({ timestamp: { $gte: cutoff } })
        .toArray();

      if (!wins.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Warzone - No Activity')
              .setDescription('No wins in the last 6 hours. The battlefield is quiet...')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const clanCounter = new Map();
      for (const w of wins) {
        if (w.clan_name) clanCounter.set(w.clan_name, (clanCounter.get(w.clan_name) || 0) + 1);
      }
      const topClans = [...clanCounter.entries()].sort((a, b) => b[1] - a[1]).slice(0, 10);

      if (!topClans.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Warzone - No Activity')
              .setDescription('No valid clan data in the last 6 hours.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const [championClan, championWins] = topClans[0];

      const clanPoints = {};
      for (const [clanName] of topClans) {
        clanPoints[clanName] = wins
          .filter((w) => w.clan_name === clanName)
          .reduce((sum, w) => sum + (w.player_count || 0), 0);
      }

      const now = new Date();
      const hoursSinceStart = now.getUTCHours() % 6;
      const minutesSinceStart = now.getUTCMinutes();
      let hoursUntilReset = 5 - hoursSinceStart;
      let minutesUntilReset = 60 - minutesSinceStart;
      if (minutesUntilReset === 60) {
        hoursUntilReset += 1;
        minutesUntilReset = 0;
      }

      const embed = new EmbedBuilder()
        .setTitle('⚔️ WARZONE - Live Battle Royale')
        .setDescription(
          `6-Hour Leaderboard | Resets in **${hoursUntilReset}h ${minutesUntilReset}m**`,
        )
        .setColor(0xff4500);

      const championPoints = clanPoints[championClan] || 0;
      embed.addFields({
        name: '👑 KING OF THE HILL',
        value: `**[${championClan}]**\n${championWins} wins | ${championPoints} points`,
        inline: false,
      });

      const rankings = topClans.map(([clan, count], i) => {
        const points = clanPoints[clan] || 0;
        let medal;
        if (i === 0) medal = '👑';
        else if (i === 1) medal = '🥈';
        else if (i === 2) medal = '🥉';
        else medal = `#${i + 1}`;
        return `${medal} [${clan}] - ${count} wins | ${points}pts`;
      });
      embed.addFields({ name: '🏆 Live Rankings', value: rankings.join('\n'), inline: false });

      const totalWins = wins.length;
      const totalPoints = wins.reduce((sum, w) => sum + (w.player_count || 0), 0);
      const contestWins = wins.filter((w) => w.contest).length;

      let statsText = `Total Battles: **${totalWins}**\n`;
      statsText += `Contest Battles: **${contestWins}**\n`;
      statsText += `Total Points: **${totalPoints}**\n`;
      statsText += `Active Clans: **${topClans.length}**`;
      embed.addFields({ name: '📊 Warzone Stats', value: statsText, inline: false });

      const latestWin = [...wins].sort(
        (a, b) => new Date(b.timestamp || 0) - new Date(a.timestamp || 0),
      )[0];
      const winTime = new Date(latestWin.timestamp || Date.now());
      const minsAgo = Math.floor((Date.now() - winTime.getTime()) / 60000);
      let timeStr;
      if (minsAgo < 1) timeStr = 'just now';
      else if (minsAgo < 60) timeStr = `${minsAgo}m ago`;
      else timeStr = `${Math.floor(minsAgo / 60)}h ago`;

      embed.addFields({
        name: '⚡ Latest Battle',
        value: `[${latestWin.clan_name || 'Unknown'}] won on ${latestWin.map || 'Unknown'} - ${timeStr}`,
        inline: false,
      });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in warzone: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to load warzone.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
